core/src/unprocessed_packet_batches.rs

```diff
@@ -1,209 +1,353 @@
+    ) -> Option<Self> {
+        let versioned_transaction: VersionedTransaction =
+            match limited_deserialize(&packet.data[0..packet.meta.size]) {
+                Ok(tx) => tx,
+                Err(_) => return None,
+            };
+
+        if let Some(message_bytes) = packet_message(&packet) {
+            let message_hash = Message::hash_raw_message(message_bytes);
+            let is_simple_vote = packet.meta.is_simple_vote_tx();
+
+            let fee_per_cu = fee_per_cu.unwrap_or_else(|| {
+                bank.as_ref()
+                    .map(|bank| compute_fee_per_cu(&versioned_transaction.message, &*bank))
+                    .unwrap_or(0)
+            });
+            Some(Self {
+                immutable_section: Rc::new(ImmutableDeserializedPacket {
+                    original_packet: packet,
+                    versioned_transaction,
+                    message_hash,
+                    is_simple_vote,
+                    fee_per_cu,
+                }),
+                forwarded: false,
+            })
+        } else {
+            None
+        }
+    }
+
+    pub fn original_packet(&self) -> &Packet {
+        &self.immutable_section.original_packet
+    }
+
+    pub fn is_simple_vote_transaction(&self) -> bool {
+        self.immutable_section.is_simple_vote
+    }
+
+    pub fn versioned_transaction(&self) -> &VersionedTransaction {
+        &self.immutable_section.versioned_transaction
+    }
+
+    pub fn sender_stake(&self) -> u64 {
+        self.immutable_section.original_packet.meta.sender_stake
+    }
+
+    pub fn message_hash(&self) -> Hash {
+        self.immutable_section.message_hash
+    }
+
+    pub fn fee_per_cu(&self) -> u64 {
+        self.immutable_section.fee_per_cu
+    }
 }
 
-/// Currently each banking_stage thread has a `UnprocessedPacketBatches` buffer to store
-/// PacketBatch's received from sigverify. Banking thread continuously scans the buffer
-/// to pick proper packets to add to the block.
-#[derive(Default)]
-pub struct UnprocessedPacketBatches(VecDeque<DeserializedPacketBatch>);
+impl PartialOrd for DeserializedPacket {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
 
-impl std::ops::Deref for UnprocessedPacketBatches {
-    type Target = VecDeque<DeserializedPacketBatch>;
-    fn deref(&self) -> &Self::Target {
-        &self.0
+impl Ord for DeserializedPacket {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.fee_per_cu().cmp(&other.fee_per_cu()) {
+            Ordering::Equal => self.sender_stake().cmp(&other.sender_stake()),
+            ordering => ordering,
+        }
     }
 }
 
-impl std::ops::DerefMut for UnprocessedPacketBatches {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.0
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+struct PacketPriorityQueueEntry {
+    fee_per_cu: u64,
+    sender_stake: u64,
+    message_hash: Hash,
```

Comment:
wdyt using `Rc<DeserializedPacket>` in place of `message_hash` to reference DeserializedPacket from priority_queue_entry? So `UnprocessedPacketBatches` could have `VecDeque<Rc<DeserializedPacket>>` instead of `HashMap`. Maybe Rust can catch it at compile time if priority_queue and stored DeserializedPacket get out of sync

Reply:
But with `message_hash` helps to prevent dups get into buffer, but dedup can be done outside

Reply:
The problem with wrapping the entire `DeserializedPacket` in an `Rc` is the `forwarded` field will no longer be mutable.